Framework/Algorithms/src/UnwrapMonitor.cpp

```diff
@@ -109,33 +109,42 @@ void UnwrapMonitor::exec() {
       // If the detector flightpath is missing, zero the data
       g_log.debug() << "Detector information for workspace index " << i
                     << " is not available.\n";
-      tempWS->mutableX(i).assign(tempWS->x(i).size(), 0.0);
-      tempWS->mutableY(i).assign(tempWS->y(i).size(), 0.0);
-      tempWS->mutableE(i).assign(tempWS->e(i).size(), 0.0);
+      tempWS->mutableX(i) = 0.0;
+      tempWS->mutableY(i) = 0.0;
+      tempWS->mutableE(i) = 0.0;
       continue;
     }
 
+    // Getting the unwrapped data is a three step process.
+    // 1. We ge the unwrapped version of the x data.
+    // 2. Then we get the unwrapped version of the y and e data for
+    //    which we require the x data from the previous step
+    // 3. Finally we need to set the newly unwrapped data on the
+    //    histogram
+
     // Unwrap the x data. Returns the bin ranges that end up being used
     const double Ld = L1 + spectrumInfo.l2(i);
     MantidVec newX;
     const std::vector<int> rangeBounds = this->unwrapX(newX, i, Ld);
+
     // Unwrap the y & e data according to the ranges found above
     MantidVec newY;
     MantidVec newE;
     this->unwrapYandE(tempWS, i, rangeBounds, newY, newE);
 
-    // Now set the new X, Y and E values on the histogram
+    // Now set the new X, Y and E values on the Histogram
     auto histogram = tempWS->histogram(i);
-    histogram.resize(newY.size());
-    auto &resizedX = histogram.mutableX();
-    auto &resizedY = histogram.mutableY();
-    auto &resizedE = histogram.mutableE();
-    std::copy(newX.begin(), newX.end(), resizedX.begin());
-    std::copy(newY.begin(), newY.end(), resizedY.begin());
-    std::copy(newE.begin(), newE.end(), resizedE.begin());
-    tempWS->setHistogram(i, histogram);
-
-    assert(tempWS->x(i).size() == tempWS->y(i).size() + 1);
+    if (histogram.yMode() == Mantid::HistogramData::Histogram::YMode::Counts) {
+        tempWS->setHistogram(
+            i, Mantid::HistogramData::BinEdges(std::move(newX)),
+            Mantid::HistogramData::Counts(std::move(newY)),
+            Mantid::HistogramData::CountStandardDeviations(std::move(newE)));
+    } else {
```

Comment:
You don't need `else`, I think. As I wrote yesterday there is a `RawCountsValidator` above, so it will always be `YMode::Counts`.